### DISCUSSION TITLE
The SSL connection could not be established - While setting up LoadBalancer using Consul and LeastConnection.

### DISCUSSION THREAD

**Discussion by ks1990cn** [ThreeMammals/Ocelot#1804] [Q&A] [answered]
Hi @raman-m @ggnaegi / Anyone ,

Can you please help me to resolve error mentioned in title? 
I've created 3 web api projects - 2 of them to replicate DownstreamApi Instance1 and Instance2, 1 is to replicate API Gateway.

when I try to reach downstream using upstream of API Gateway, and when it tries to find hostandport anc check for services available, it suddenly breaks stating SSL Connection could not be established.

Can you help me by cloning below code and let me know what I am missing in setup. Here is GIF of problem and code from where you can clone.

Please let me know any input required from me.

![SSLissue](https://github.com/ThreeMammals/Ocelot/assets/29522704/8d8b9bc8-840a-4806-851f-29695ed6b572)

Code path changes I've done:
https://github.com/ks1990cn/Ocelot/tree/LoadBalancerSample

**Comment by raman-m** [comment]
Tanmay, I'm busy with
- #1805 

Regarding SSL errors... Show us your routes JSON please!
Which certificate do you use: SSL signing authority or self-signed certificate?

Did you read this: [SSL Errors](https://ocelot.readthedocs.io/en/latest/features/configuration.html#ssl-errors) ❓

**Reply by ks1990cn** [reply]
Here you go!

Just added DangerousAcceptAnyServerCertificateValidator : true but not worked for me, I am trying to run on http.
```
{
  "Routes": [
    {
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 7156
        },
        {
          "Host": "localhost",
          "Port": 7084
        }
      ],
      "ServiceName": "ConsulDemo",
      "LoadBalancerOptions": {
        "Type": "LeastConnection"
      },
      "UpstreamPathTemplate": "/hostname",
      "UpstreamHttpMethod": [ "GET" ],
      "DownstreamPathTemplate": "/WeatherForecast/hostname",
      "DownstreamScheme": "https",
      "DangerousAcceptAnyServerCertificateValidator": true


    }
  ],
  "GlobalConfiguration": {
    "ServiceDiscoveryProvider": {
      "Scheme": "https",
      "Host": "localhost",
      "Port": 8500,
      "Type": "Consul"
    }
  }
}

```

**Reply by ks1990cn** [reply]
Tried with http also, but faced same here is my route:
```

{
  "Routes": [
    {
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 5022
        },
        {
          "Host": "localhost",
          "Port": 5085
        }
      ],
      "ServiceName": "ConsulDemo",
      "LoadBalancerOptions": {
        "Type": "LeastConnection"
      },
      "UpstreamPathTemplate": "/hostname",
      "UpstreamHttpMethod": [ "GET" ],
      "DownstreamPathTemplate": "/WeatherForecast/hostname",
      "DownstreamScheme": "http",
      "DangerousAcceptAnyServerCertificateValidator": true


    }
  ],
  "GlobalConfiguration": {
    "ServiceDiscoveryProvider": {
      "Scheme": "https",
      "Host": "localhost",
      "Port": 8500,
      "Type": "Consul"
    }
  }
}
```

In Program.cs of API Gateway, I have added following code:

```
builder.Configuration.AddJsonFile("Ocelot.json");
builder.Services.AddOcelot(builder.Configuration).AddConsul();
app.UseOcelot().Wait();
```

**Reply by ggnaegi** [reply]
@ks1990cn you should check the consul parameters. https://developer.hashicorp.com/consul/docs/install/ports 8500 is port for http scheme.

**Reply by raman-m** [reply]
Moreover, Consul server default port is configurable. And this def port is open for connections.
So, Tanmay have to check Consul settings port, and Ocelot "ServiceDiscoveryProvider" port.
That's why a connection cannot be established. This is my understanding...

**Reply by ks1990cn** [reply]
Let me try again

**Reply by ks1990cn** [reply, marked as answer]
Its fixed now, I hosted consul on right port, my scheme was https instead http, and it was causing not to run. Thanks alot team!!

 ```
"ServiceDiscoveryProvider": {
      "Scheme": "https",
      "Host": "localhost",
      "Port": 8500,
      "Type": "Consul"
    }
```
<img width="274" alt="image" src="https://github.com/ThreeMammals/Ocelot/assets/29522704/c21aed30-c425-4218-8232-bb6f3dc7f60e">

**Reply by raman-m** [reply]
Congrats on fixing it finally! :partying_face:  
Could we close this discussion plz?

**Reply by raman-m** [reply]
If you use Docker containers then it seems you use self-signed certificates.
There is no sense to use `https` scheme for local development environment! So, yes, switching to `http` scheme should solve any SSL connection problems.

**Reply by ks1990cn** [reply]
Correct close this issue,
I've added these changes under sample folder of Ocelot in my fork.

What you guys think, should I publish this and recorded GIF (for documentation) so it can be sample setup for other contributors?

Do you want these type of contributions or just prefer
Bug work for now?

@raman-m @ggnaegi

**Reply by ks1990cn** [reply]
I can raise PR and documentations which you can approve.

**Reply by raman-m** [reply]
Did you create an issue?

**Reply by ks1990cn** [reply]
No, I just raised PR for it. Proposing a new sample to replicate issues quickly.

Should I raise issue for it first?

**Reply by raman-m** [reply]
In general, we create issues first.
I will create an issue... letting you to focus on PR code...
